### DISCUSSION TITLE
Reduced Regression Models for Time Series Forecasting

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1760] [FAQs]
# Background

Usually, we can not use a regression model directly to build a forecasting model for data with temporal dependence (fancy term for time series data). In order to do so, we need to first convert the temporal data into a supervised learning problem with appropriate features. For the sake of this discussion, we will call his "reduced regression". More details on this topic can be [found here](https://machinelearningmastery.com/convert-time-series-supervised-learning-problem-python/).

An important concern during this process is to make sure that we do not leak information about any future (unavailable) data while performing this conversion (see [Pitfall 1 here](https://nbviewer.org/github/alan-turing-institute/sktime/blob/v0.5.3/examples/01_forecasting.ipynb)).

Another important concern here is the ability to do multi-step ahead forecasting using this approach. Since the data is formulated as a single step ahead supervised problem, performing multi-step ahead forecasting needs addition work which the modeler must take care of (see [Pitfall 3 here](https://nbviewer.org/github/alan-turing-institute/sktime/blob/v0.5.3/examples/01_forecasting.ipynb)).

# Solution

The `sktime` library provides a very convenient way to perform Time Series Forecasting using Reduced Regression models and `pycaret` has incorporated the same into its time series module as well. However, there are some nuances that need to be taken into consideration when using these models. For example, since the Reduced Regression models do not have any notion of time, trends, and seasonality, it is often useful to deseasonalize and de-trend the data before performing the reduction, training and forecasting. This is analogous to modeling only the auto-regressive terms in the data which makes it easier to find a better fitting model since we are not relying on it to figure out seasonality and trend in addition to the auto-regressive properties. 

This is taken care of  automatically in `pycaret` using a pipeline which conditionally deseasonalizes (if seasonality is detected) and de-trends the data before fitting it. These models are represented with a suffix `_cds_dt`. The time series module currently supports almost 20 reduced regression models of various types (linear, decision trees, gradient boosted, nearest neighbors, etc).

```python
#### Install library (one time) ----
!pip install pycaret-ts-alpha -U

#### Load data ----
from pycaret.datasets import get_data
y = get_data("lynx")

#### Setup experiment ----
from pycaret.internal.pycaret_experiment import TimeSeriesExperiment
exp = TimeSeriesExperiment()
exp.setup(data=y, fh=11, seasonal_period=11, session_id=42)  # Note that we provided seasonal_period manually for annual data

##### List the available models ----
exp.models()
```
![image](https://user-images.githubusercontent.com/33585645/138571829-4e470d7f-fcdf-4704-8f2d-affcf84b9b81.png)

```python
#### Train a Reduced Regression Forecaster that uses a Linear Regression Model ----
model = exp.create_model("lr_cds_dt")

#### Plot the model predictons ----
exp.plot_model(model)
```

![image](https://user-images.githubusercontent.com/33585645/138572053-31ab4599-ae52-47b4-bbac-27caca333c54.png)

**Note how we are able to take care of all the pitfalls above (reduction of data, preventing leakage, multi-step ahead forecasting) with a single line of code!** Now that is convenience at your fingertips.

That's it for this FAQ. Happy forecasting!

# Suggested Next Reads

- #1848

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
